fjord/heartbeat/healthchecks.py

```diff
@@ -0,0 +1,205 @@
+            cls.name,
+            SEVERITY_LOW,
+            '%s answers in last 24 hours.' % str(count),
+            str(count)
+        )
+
+
+def tableify(table):
+    num_cols = 0
+    maxes = []
+
+    for row in table:
+        num_cols = max(num_cols, len(row))
+        if len(maxes) < len(row):
+            maxes.extend([0] * (len(row) - len(maxes)))
+        for i, cell in enumerate(row):
+            maxes[i] = max(maxes[i], len(str(cell)))
+
+    def fix_row(maxes, row):
+        return '  '.join([
+            str(cell) + (' ' * (maxes[i] - len(str(cell))))
+            for i, cell in enumerate(row)
+        ])
+
+    return '\n'.join(
+        [
+            fix_row(maxes, row)
+            for row in table
+        ]
+    )
+
+
+@register_check
+class CheckMissingVotes(Check):
+    """should freak out if votes are 0 for any 'large' cell."""
+    name = 'missing votes check'
+
+    @classmethod
+    def check(cls):
+        # Note: This SQL statement comes from Gregg. It's probably
+        # mysql-specific.
+        sql = """
+        SELECT
+            sum(score is not NULL) as nvoted,
+            DATE_FORMAT(received_ts, '%Y-%m-%d') as ydm,
+            version,
+            channel,
+            100*sum(flow_began_ts > 0) / count(received_ts) as pct_began,
+            100*sum(flow_offered_ts >0) / count(received_ts) as pct_offered,
+            100*sum(flow_voted_ts > 0)/ count(received_ts) as pct_voted,
+            100*sum(flow_engaged_ts > 0) / count(received_ts) as pct_engaged,
+            count(received_ts) as N
+        FROM heartbeat_answer
+        WHERE
+            received_ts > DATE_SUB(now(), interval 1 day)
+            AND is_test=0
+            AND survey_id="heartbeat-by-user-first-impression"
+            AND (locale='en-us')
+        GROUP BY version, channel, ydm
+        HAVING
+            N >= 50
+            and nvoted = 0
+        ORDER BY
+            channel,
+            version,
+            ydm;
+        """
+        cursor = connection.cursor()
+        cursor.execute(sql)
+        data = cursor.fetchall()
+
+        if not data:
+            # If we get nothing back, then we have serious issues.
+            return Result(
+                cls.name,
+                SEVERITY_HIGH,
+                'No data from query',
+                repr(data)
+            )
+
+        data = list(data)
+
+        # FIXME: What consistutes SEVERITY_HIGH here?
+
+        data.insert(
+            0,
+            ['nvoted', 'ydm', 'version', 'channel', 'pct_began', 'pct_offered',
+             'pct_voted', 'pct_engaged', 'N']
+        )
+        return Result(
+            cls.name,
+            SEVERITY_LOW,
+            'Data looks ok.',
+            tableify(data)
+        )
+
+
+def get_all_healthchecks():
+    return CHECKS
+
+
+def run_healthchecks():
+    return [checker.check() for checker in get_all_healthchecks()]
+
+
+def email_healthchecks(results):
+    counter = Counter()
```

Comment:
It's not a set... it's counting the number of HIGH severity results there are. I could do something like:

```
has_high = any([result == SEVERITY_HIGH for result in results])
```

That'd probably work for the things I'm using it for now.

I decided to use a counter in case we wanted to add a summary to the top of the email like:

```
HIGH: 5 checks
MEDIUM: 0 checks
LOW: 2 checks
```

Or maybe in the subject line:

```
H: 5 M: 0 L: 2
```

I don't know how many checks there will be.

Also, one of the requirements is that Gregg can skim it quickly. So the output is mostly affected by what he can absorb in a quick skim.

That's what was in my head at the time. For now, doing an `has_high` is fine.